README.md
# daixijun.mysql

[![Build Status](https://github.com/daixijun/ansible-role-mysql/workflows/build/badge.svg)](https://github.com/daixijun/ansible-role-mysql/actions)
[![Ansible Galaxy](https://img.shields.io/badge/galaxy-daixijun.mysql-660198.svg?style=flat)](https://galaxy.ansible.com/daixijun/mysql/)
[![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/daixijun/ansible-role-mysql?sort=semver)](https://github.com/daixijun/ansible-role-mysql/tags)

用于快速部署 mysql 集群

支持以下几种集群模式:

* [x] Master-Slave 基于GTID/binlog的主从复制
* [x] MGR 单主模式(默认)
* [x] MGR 多主模式

## 环境要求

* Centos 7+
* Ansible 2.9+
* MySQL 8.0+

## 变量

| 变量名                                      | 类型       | 默认值                      | 变量说明                                                                                                                                                              |
| ------------------------------------------- | ---------- | --------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| mysql_version                               | string     | 8.0.20                      | mysql 版本                                                                                                                                                            |
| mysql_download_url                          | string     |                             | 免安装压缩包下载地址                                                                                                                                                  |
| mysql_inventory_group                       | string     | mysql                       | Ansible 主机清单中的组名， 默认为 `mysql`， 其对应的 primary 与 secondary 组名分别为 `{{ mysql_inventory_group }}_primary` 与 `{{ mysql_inventory_group }}_secondary` |
| mysql_datadir                               | string     | /data/mysql                 |                                                                                                                                                                       | 数据存放目录                               |
| mysql_logdir                                | string     | /var/log/mysqld             | 日志存放目录                                                                                                                                                          |
| mysql_pidfile                               | string     | /var/run/mysqld/mysqld.pid  | PID文件位置                                                                                                                                                           |
| mysql_socket                                | string     | /var/run/mysqld/mysqld.sock | Socket文件位置                                                                                                                                                        |
| mysql_port                                  | int        | 3306                        | 监听端口                                                                                                                                                              |
| mysql_bind_address                          | string     | 0.0.0.0                     | 监听地址                                                                                                                                                              |
| mysql_interface                             | string     |                             | 指定网卡，默认使用除lo外的第一张网卡                                                                                                                                  |
| mysql_default_time_zone                     | string     | +8:00                       | 指定时区                                                                                                                                                              |
| mysql_character_set_server                  | string     | utf8mb4                     | 默认字符集                                                                                                                                                            |
| mysql_collation_server                      | string     | utf8mb4_general_ci          | 默认字符序                                                                                                                                                            |
| mysql_max_connections                       | string/int | 1005                        | 最大连接数                                                                                                                                                            |
| mysql_max_user_connections                  | string/int | 1000                        | 用户最大连接数，必须比 `mysql_max_connections` 小，需要给管理员预留几个连接用于处理异常情况                                                                           |
| mysql_max_connect_errors                    | string/int | 200                         | 最大错误连接数                                                                                                                                                        |
| mysql_root_password                         | string     |                             | root账号的密码                                                                                                                                                        |
| mysql_cluster_type                          | string     | mgr                         | 集群类型(默认 mgr) 可选 `mgr`(Mysql Group Replication)/`ms`(Master-Slave)                                                                                             |
| mysql_cluster_role                          | string     | primary                     | Primary-Secondary 模式下的实例角色，可选                                                                                                                              | `master`, `primary` / `slave`, `secondary` |
| mysql_replication_master                    | string     |                             | Master-Slave 模式下Master实例的名称                                                                                                                                   |
| mysql_replication_based                     | string     | gtid                        | 可选基于 `gtid` 或传统 `binlog` 方式进行复制(默认 gtid)                                                                                                               |
| mysql_repl_user                             | string     | repl                        | 用于主从/组复制的账号                                                                                                                                                 |
| mysql_repl_password                         | string     |                             | 用于主从/组复制的账号的密码                                                                                                                                           |
| mysql_group_replication_name                | uuid       |                             | 组复制集群名,在mysql中使用`SELECT UUID()`或shell中使用`uuidgen`生成                                                                                                   |
| mysql_group_replication_single_primary_mode | bool       | true                        | MGR集群是否为单主模式                                                                                                                                                 |
| mysql_innodb_cluster_enable                 | bool       | true                        | 是否开启 Innodb Cluster                                                                                                                                               |
| mysql_innodb_cluster_name                   | string     | default                     | Innodb Cluster 名称                                                                                                                                                   |
| mysql_innodb_cluster_username               | string     | ic                          | 用于创建和管理 Innodb Cluster 的账号，需要具备 `ALL WITH GRANT OPTION` 权限                                                                                           |
| mysql_innodb_cluster_password               | string     | ""                          | 管理密码                                                                                                                                                              |
| mysql_proxysql_config                       | bool       | false                       | 是否使用 proysql 做为代理层                                                                                                                                           |
| mysql_proxysql_monitor_username             | string     | monitor                     | proxysql 监控mgr状态的账号                                                                                                                                            |
| mysql_proxysql_monitor_password             | string     | ""                          | proxysql 监控mgr状态的密码                                                                                                                                            |
| mysql_databases                             | array      | []                          | 需要创建的业务数据库                                                                                                                                                  |
| mysql_users                                 | array      | []                          | 需要创建的用户                                                                                                                                                        |

## 依赖

collections:

* [community.general](https://github.com/ansible-collections/community.general)

## 示例

安装

```shell
ansible-galaxy collection install -r requirements.yml
ansible-galaxy install daixijun.mysql
```

使用

```yaml
- hosts: servers
  collections:
    - community.general
  roles:
    - { role: daixijun.mysql, mysql_version: 8.0.20 }
```

## License

BSD

## 已知问题

* [ ] mysql_user 模块对于mysql8.0以上的版本，给用户授权`ALL`权限的时候会出现幂等性问题 [Idempotence all grant](https://github.com/ansible/ansible/pull/57460)
* [ ] 目前配置文件有改动时，重启mysql服务后可能会造成集群异常

## 待办

* [ ] 主从模式下支持半同步复制

## 维护者

* Xijun Dai <daixijun1990@gmail.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.3.3](https://github.com/daixijun/ansible-role-mysql/compare/v2.3.2...v2.3.3) (2020-07-02)


### Bug Fixes

* 修复 delegate_to 为 localhost 时可能会被处于指定组外的主机问题 ([6731f67](https://github.com/daixijun/ansible-role-mysql/commit/6731f671cf4c0314b5766c7a6cc0251fa45e5b15))
* 添加 mysql_shell_download_url 变量用于自定义 rpm 包下载地址 ([4604875](https://github.com/daixijun/ansible-role-mysql/commit/4604875898d201f2681710d6fc21ff791003dfa4))

### [2.3.2](https://github.com/daixijun/ansible-role-mysql/compare/v2.3.1...v2.3.2) (2020-07-02)


### Bug Fixes

* 查询主节点时使用委托方式，解决 play 中主机不完全用于mysql时会造成跳过的情况 ([27ac753](https://github.com/daixijun/ansible-role-mysql/commit/27ac75385042c3f042b14f122fce47e9fea3b8bf))

### [2.3.1](https://github.com/daixijun/ansible-role-mysql/compare/v2.3.0...v2.3.1) (2020-06-30)


### Bug Fixes

* 修改变量检测逻辑 ([b40baef](https://github.com/daixijun/ansible-role-mysql/commit/b40baefe2763f3a596b0819936f2ae67dd0d2674))

## [2.3.0](https://github.com/daixijun/ansible-role-mysql/compare/v2.2.1...v2.3.0) (2020-06-28)


### Features

* 创建db时可以通过 datafile 参数指定需要导入的sql文件, 仅在db为空时导入 ([1a4c0af](https://github.com/daixijun/ansible-role-mysql/commit/1a4c0af01d00609b8fd9bcdcafca4ee92bb6ae4b))
* 添加 mysql_inventory_group 参数用于指定 inventory 中的组名 ([29d8cbc](https://github.com/daixijun/ansible-role-mysql/commit/29d8cbc038c968355425d54aad39cbffe79e921e))
* 添加 mysql_port 与 mysql_bind_address 参数用于自定义端口及监听 ([1ec2822](https://github.com/daixijun/ansible-role-mysql/commit/1ec282298e05569f1cff300f2be40271a24fba36))
* 移除 library, 依赖于 collection ([c6402a9](https://github.com/daixijun/ansible-role-mysql/commit/c6402a9c25c492b4538481ac306ad2dcd38cf938))
* **proxysql:** 添加 proxysql 相关配置 ([0210219](https://github.com/daixijun/ansible-role-mysql/commit/0210219f2f2765e4fa722d3c4b83768bd8d34ee6))


### Bug Fixes

* 使用 groups['mysql'] 代替 ansible_play_hosts 变量 ([f135714](https://github.com/daixijun/ansible-role-mysql/commit/f13571409137047bf4a1ddcc4638e55e5ae3fb18))
* 修复 data file 加载逻辑 ([2dcafc9](https://github.com/daixijun/ansible-role-mysql/commit/2dcafc9bc3508914e0f93ff7cad94ee1ea9a07eb))
* 修复 datafile 判断 ([ed628b1](https://github.com/daixijun/ansible-role-mysql/commit/ed628b1662f62faf14419615ca24e2046df0b09d))
* 修复 mysql_cluster_role 未支持 secondary 的问题 ([d77c483](https://github.com/daixijun/ansible-role-mysql/commit/d77c4835fb9bf4a7f747fad7db341e2014ae747d))
* 修复 mysql-shell 下载安装包时取消become ([79e8f08](https://github.com/daixijun/ansible-role-mysql/commit/79e8f08d3658f8fc4f20e1bfb23144903280c3e4))
* 修复组复制中 mysql_repl_user 变量未使用的问题 ([13bad35](https://github.com/daixijun/ansible-role-mysql/commit/13bad358e6b124c14716a57c6c291610fbc3201c))
* 修改 sys.gr_member_routing_candidate_status 判断逻辑 ([1ca3329](https://github.com/daixijun/ansible-role-mysql/commit/1ca3329cf8aa4d19a4f8771e7f2c409fe7c8be07))
* 创建db时支持设置自定义字符集 ([ffa2050](https://github.com/daixijun/ansible-role-mysql/commit/ffa2050fe445de4361396723818bae1fae09cd26))
* 添加 mysqladmin 相关认证配置项 ([2deeef1](https://github.com/daixijun/ansible-role-mysql/commit/2deeef163efc9f3be56105849fd727ad42882590))
* 确保组复制名称为小写uuid值 ([0b47f06](https://github.com/daixijun/ansible-role-mysql/commit/0b47f062297ddcd49343e348b1beb021e1cfa1b2))

### [2.2.1](https://github.com/daixijun/ansible-role-mysql/compare/v2.2.0...v2.2.1) (2020-06-22)


### Bug Fixes

* 修复 vault 后 length 变量判断失败错误 ([6a6a092](https://github.com/daixijun/ansible-role-mysql/commit/6a6a09232a81a6655070f3cb0b79017ffc22f026))

## [2.2.0](https://github.com/daixijun/ansible-role-mysql/compare/v2.1.2...v2.2.0) (2020-06-22)


### Features

* mgr集群支持开启Innodb Cluster功能 ([abcf5d5](https://github.com/daixijun/ansible-role-mysql/commit/abcf5d51032b21b2d3839f66b45d988ae24d419e))

### [2.1.2](https://github.com/daixijun/ansible-role-mysql/compare/v2.1.1...v2.1.2) (2020-06-19)


### Bug Fixes

* 修复when中多出的引号问题 ([726b287](https://github.com/daixijun/ansible-role-mysql/commit/726b2879ebfbf3269162fbe0d2aa9e240c82465d))
* 修改 master-slave 为 primary-secondary ([632b0ba](https://github.com/daixijun/ansible-role-mysql/commit/632b0ba2ee9bbba9317a5267f8c2ef351a7cdc94))
* 移除默认密码 ([7cac1db](https://github.com/daixijun/ansible-role-mysql/commit/7cac1db9013ed7e454e7882c0163346446938431))
* **defaults:** 更新默认版本为 8.0.20 ([58ce130](https://github.com/daixijun/ansible-role-mysql/commit/58ce13023aa11b44b20f9ce8385cb41610ad3d62))
* 添加默认 collation_server 参数 ([02595f0](https://github.com/daixijun/ansible-role-mysql/commit/02595f0cf7114a3c8771a63fdc671550b02a8ffc))

### [2.1.1](https://github.com/daixijun/ansible-role-mysql/compare/v2.1.0...v2.1.1) (2020-04-28)


### Bug Fixes

* 修复归档版本下载地址 ([61eaa00](https://github.com/daixijun/ansible-role-mysql/commit/61eaa00cf360abd402dc13f8cffa4f5c50db7cd9))

## [2.1.0](https://github.com/daixijun/ansible-role-mysql/compare/v2.0.0...v2.1.0) (2020-04-26)


### Features

* 主从复制支持通过 mysql_replication_based 变量配置为基于 gtid 或 binlog 方式同步 ([d4debc6](https://github.com/daixijun/ansible-role-mysql/commit/d4debc6c806b1965ed4a3bae48fdd2a605fd8693))
* 添加 mysql_replication_based 变量 ([d070eec](https://github.com/daixijun/ansible-role-mysql/commit/d070eec063f421bf88dc056d345f202628d85577))

## [2.0.0](https://github.com/daixijun/ansible-role-mysql/compare/v1.5.2...v2.0.0) (2020-04-22)


### ⚠ BREAKING CHANGES

* 修改默认sql_mode为空
* 修改 Master-Slave 同步试为基于GTID的主从复制

### Features

* 修改 Master-Slave 为基于GTID方式进行主从复制 ([278049e](https://github.com/daixijun/ansible-role-mysql/commit/278049ec1a284df418d10395e6f547b77ec2e685))
* 增加 mysql 依赖包对 RHEL/CentOS 7/8 的支持 ([2ad8426](https://github.com/daixijun/ansible-role-mysql/commit/2ad8426a03bd56dbd5b6ae606dddf4dfa78a4654))
* 添加 ~/.my.cnf 模板文件 ([b8ab69e](https://github.com/daixijun/ansible-role-mysql/commit/b8ab69e81f137faf1a2432cd09dadb9bdfe486b2))

molecule/master_slave/tests/test_default.py
import os

import testinfra.utils.ansible_runner

testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(os.environ["MOLECULE_INVENTORY_FILE"]).get_hosts("all")


def test_mysqld_runing(host):
    mysql = host.service("mysqld")

    assert mysql.is_running


library/mysql_replication.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2013, Balazs Pocze <banyek@gawker.com>
# Certain parts are taken from Mark Theunissen's mysqldb module
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


ANSIBLE_METADATA = {"metadata_version": "1.1", "status": ["preview"], "supported_by": "community"}


DOCUMENTATION = r"""
---
module: mysql_replication
short_description: Manage MySQL replication
description:
    - Manages MySQL server replication, slave, master status get and change master host.
version_added: "1.3"
author:
- Balazs Pocze (@banyek)
options:
    mode:
        description:
            - module operating mode. Could be getslave (SHOW SLAVE STATUS), getmaster (SHOW MASTER STATUS), changemaster (CHANGE MASTER TO), startslave
              (START SLAVE), stopslave (STOP SLAVE), resetslave (RESET SLAVE), resetslaveall (RESET SLAVE ALL)
        type: str
        choices:
            - getslave
            - getmaster
            - changemaster
            - stopslave
            - startslave
            - resetslave
            - resetslaveall
        default: getslave
    master_host:
        description:
            - Same as mysql variable.
        type: str
    master_user:
        description:
            - Same as mysql variable.
        type: str
    master_password:
        description:
            - Same as mysql variable.
        type: str
    master_port:
        description:
            - Same as mysql variable.
        type: int
    master_connect_retry:
        description:
            - Same as mysql variable.
        type: int
    master_log_file:
        description:
            - Same as mysql variable.
        type: str
    master_log_pos:
        description:
            - Same as mysql variable.
        type: int
    relay_log_file:
        description:
            - Same as mysql variable.
        type: str
    relay_log_pos:
        description:
            - Same as mysql variable.
        type: int
    master_ssl:
        description:
            - Same as mysql variable.
        type: bool
    master_ssl_ca:
        description:
            - Same as mysql variable.
        type: str
    master_ssl_capath:
        description:
            - Same as mysql variable.
        type: str
    master_ssl_cert:
        description:
            - Same as mysql variable.
        type: str
    master_ssl_key:
        description:
            - Same as mysql variable.
        type: str
    master_ssl_cipher:
        description:
            - Same as mysql variable.
        type: str
    master_auto_position:
        description:
            - Whether the host uses GTID based replication or not.
        type: bool
        version_added: "2.0"
    master_use_gtid:
        description:
            - Configures the slave to use the MariaDB Global Transaction ID.
            - C(disabled) equals MASTER_USE_GTID=no command.
            - To find information about available values see
              U(https://mariadb.com/kb/en/library/change-master-to/#master_use_gtid).
            - Available since MariaDB 10.0.2.
        choices: [current_pos, slave_pos, disabled]
        type: str
        version_added: "2.10"
    master_delay:
        description:
            - Time lag behind the master's state (in seconds).
            - Available from MySQL 5.6.
            - For more information see U(https://dev.mysql.com/doc/refman/8.0/en/replication-delayed.html).
        type: int
        version_added: "2.10"
    connection_name:
        description:
            - Name of the master connection.
            - Supported from MariaDB 10.0.1.
            - For more information see U(https://mariadb.com/kb/en/library/multi-source-replication/).
        type: str
        version_added: "2.10"
    channel:
        description:
            - Name of replication channel.
            - Multi-source replication is supported from MySQL 5.7.
            - For more information see U(https://dev.mysql.com/doc/refman/8.0/en/replication-multi-source.html).
        type: str
        version_added: "2.10"

extends_documentation_fragment:
- mysql
"""

EXAMPLES = r"""
- name: Stop mysql slave thread
  mysql_replication:
    mode: stopslave

- name: Get master binlog file name and binlog position
  mysql_replication:
    mode: getmaster

- name: Change master to master server 192.0.2.1 and use binary log 'mysql-bin.000009' with position 4578
  mysql_replication:
    mode: changemaster
    master_host: 192.0.2.1
    master_log_file: mysql-bin.000009
    master_log_pos: 4578

- name: Check slave status using port 3308
  mysql_replication:
    mode: getslave
    login_host: ansible.example.com
    login_port: 3308

- name: On MariaDB change master to use GTID current_pos
  mysql_replication:
    mode: changemaster
    master_use_gtid: current_pos

- name: Change master to use replication delay 3600 seconds
  mysql_replication:
    mode: changemaster
    master_host: 192.0.2.1
    master_delay: 3600

- name: Start MariaDB standby with connection name master-1
  mysql_replication:
    mode: startslave
    connection_name: master-1

- name: Stop replication in channel master-1
  mysql_replication:
    mode: stopslave
    channel: master-1
"""

RETURN = r"""
queries:
  description: List of executed queries which modified DB's state.
  returned: always
  type: list
  sample: ["CHANGE MASTER TO MASTER_HOST='master2.example.com',MASTER_PORT=3306"]
  version_added: '2.10'
"""

import os
import warnings

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.mysql import mysql_connect, mysql_driver, mysql_driver_fail_msg
from ansible.module_utils._text import to_native

executed_queries = []


def get_master_status(cursor):
    cursor.execute("SHOW MASTER STATUS")
    masterstatus = cursor.fetchone()
    return masterstatus


def get_slave_status(cursor, connection_name="", channel=""):
    if connection_name:
        query = "SHOW SLAVE '%s' STATUS" % connection_name
    else:
        query = "SHOW SLAVE STATUS"

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    cursor.execute(query)
    slavestatus = cursor.fetchone()
    return slavestatus


def stop_slave(cursor, connection_name="", channel=""):
    if connection_name:
        query = "STOP SLAVE '%s'" % connection_name
    else:
        query = "STOP SLAVE"

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    try:
        executed_queries.append(query)
        cursor.execute(query)
        stopped = True
    except Exception:
        stopped = False
    return stopped


def reset_slave(cursor, connection_name="", channel=""):
    if connection_name:
        query = "RESET SLAVE '%s'" % connection_name
    else:
        query = "RESET SLAVE"

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    try:
        executed_queries.append(query)
        cursor.execute(query)
        reset = True
    except Exception:
        reset = False
    return reset


def reset_slave_all(cursor, connection_name="", channel=""):
    if connection_name:
        query = "RESET SLAVE '%s' ALL" % connection_name
    else:
        query = "RESET SLAVE ALL"

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    try:
        executed_queries.append(query)
        cursor.execute(query)
        reset = True
    except Exception:
        reset = False
    return reset


def start_slave(cursor, connection_name="", channel=""):
    if connection_name:
        query = "START SLAVE '%s'" % connection_name
    else:
        query = "START SLAVE"

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    try:
        executed_queries.append(query)
        cursor.execute(query)
        started = True
    except Exception:
        started = False
    return started


def changemaster(cursor, chm, connection_name="", channel=""):
    if connection_name:
        query = "CHANGE MASTER '%s' TO %s" % (connection_name, ",".join(chm))
    else:
        query = "CHANGE MASTER TO %s" % ",".join(chm)

    if channel:
        query += " FOR CHANNEL '%s'" % channel

    executed_queries.append(query)
    cursor.execute(query)


def main():
    module = AnsibleModule(
        argument_spec=dict(
            login_user=dict(type="str"),
            login_password=dict(type="str", no_log=True),
            login_host=dict(type="str", default="localhost"),
            login_port=dict(type="int", default=3306),
            login_unix_socket=dict(type="str"),
            mode=dict(
                type="str", default="getslave", choices=["getmaster", "getslave", "changemaster", "stopslave", "startslave", "resetslave", "resetslaveall"]
            ),
            master_auto_position=dict(type="bool", default=False),
            master_host=dict(type="str"),
            master_user=dict(type="str"),
            master_password=dict(type="str", no_log=True),
            master_port=dict(type="int"),
            master_connect_retry=dict(type="int"),
            master_log_file=dict(type="str"),
            master_log_pos=dict(type="int"),
            relay_log_file=dict(type="str"),
            relay_log_pos=dict(type="int"),
            master_ssl=dict(type="bool", default=False),
            master_ssl_ca=dict(type="str"),
            master_ssl_capath=dict(type="str"),
            master_ssl_cert=dict(type="str"),
            master_ssl_key=dict(type="str"),
            master_ssl_cipher=dict(type="str"),
            connect_timeout=dict(type="int", default=30),
            config_file=dict(type="path", default="~/.my.cnf"),
            client_cert=dict(type="path", aliases=["ssl_cert"]),
            client_key=dict(type="path", aliases=["ssl_key"]),
            ca_cert=dict(type="path", aliases=["ssl_ca"]),
            master_use_gtid=dict(type="str", choices=["current_pos", "slave_pos", "disabled"]),
            master_delay=dict(type="int"),
            connection_name=dict(type="str"),
            channel=dict(type="str"),
        )
    )
    mode = module.params["mode"]
    master_host = module.params["master_host"]
    master_user = module.params["master_user"]
    master_password = module.params["master_password"]
    master_port = module.params["master_port"]
    master_connect_retry = module.params["master_connect_retry"]
    master_log_file = module.params["master_log_file"]
    master_log_pos = module.params["master_log_pos"]
    relay_log_file = module.params["relay_log_file"]
    relay_log_pos = module.params["relay_log_pos"]
    master_ssl = module.params["master_ssl"]
    master_ssl_ca = module.params["master_ssl_ca"]
    master_ssl_capath = module.params["master_ssl_capath"]
    master_ssl_cert = module.params["master_ssl_cert"]
    master_ssl_key = module.params["master_ssl_key"]
    master_ssl_cipher = module.params["master_ssl_cipher"]
    master_auto_position = module.params["master_auto_position"]
    ssl_cert = module.params["client_cert"]
    ssl_key = module.params["client_key"]
    ssl_ca = module.params["ca_cert"]
    connect_timeout = module.params["connect_timeout"]
    config_file = module.params["config_file"]
    master_delay = module.params["master_delay"]
    if module.params.get("master_use_gtid") == "disabled":
        master_use_gtid = "no"
    else:
        master_use_gtid = module.params["master_use_gtid"]
    connection_name = module.params["connection_name"]
    channel = module.params["channel"]

    if mysql_driver is None:
        module.fail_json(msg=mysql_driver_fail_msg)
    else:
        warnings.filterwarnings("error", category=mysql_driver.Warning)

    login_password = module.params["login_password"]
    login_user = module.params["login_user"]

    try:
        cursor = mysql_connect(
            module, login_user, login_password, config_file, ssl_cert, ssl_key, ssl_ca, None, cursor_class="DictCursor", connect_timeout=connect_timeout
        )
    except Exception as e:
        if os.path.exists(config_file):
            module.fail_json(
                msg="unable to connect to database, check login_user and login_password are correct or %s has the credentials. "
                "Exception message: %s" % (config_file, to_native(e))
            )
        else:
            module.fail_json(msg="unable to find %s. Exception message: %s" % (config_file, to_native(e)))

    if mode in "getmaster":
        status = get_master_status(cursor)
        if not isinstance(status, dict):
            status = dict(Is_Master=False, msg="Server is not configured as mysql master")
        else:
            status["Is_Master"] = True
        module.exit_json(queries=executed_queries, **status)

    elif mode in "getslave":
        status = get_slave_status(cursor, connection_name, channel)
        if not isinstance(status, dict):
            status = dict(Is_Slave=False, msg="Server is not configured as mysql slave")
        else:
            status["Is_Slave"] = True
        module.exit_json(queries=executed_queries, **status)

    elif mode in "changemaster":
        chm = []
        result = {}
        if master_host:
            chm.append("MASTER_HOST='%s'" % master_host)
        if master_user:
            chm.append("MASTER_USER='%s'" % master_user)
        if master_password:
            chm.append("MASTER_PASSWORD='%s'" % master_password)
        if master_port is not None:
            chm.append("MASTER_PORT=%s" % master_port)
        if master_connect_retry is not None:
            chm.append("MASTER_CONNECT_RETRY=%s" % master_connect_retry)
        if master_log_file:
            chm.append("MASTER_LOG_FILE='%s'" % master_log_file)
        if master_log_pos is not None:
            chm.append("MASTER_LOG_POS=%s" % master_log_pos)
        if master_delay is not None:
            chm.append("MASTER_DELAY=%s" % master_delay)
        if relay_log_file:
            chm.append("RELAY_LOG_FILE='%s'" % relay_log_file)
        if relay_log_pos is not None:
            chm.append("RELAY_LOG_POS=%s" % relay_log_pos)
        if master_ssl:
            chm.append("MASTER_SSL=1")
        if master_ssl_ca:
            chm.append("MASTER_SSL_CA='%s'" % master_ssl_ca)
        if master_ssl_capath:
            chm.append("MASTER_SSL_CAPATH='%s'" % master_ssl_capath)
        if master_ssl_cert:
            chm.append("MASTER_SSL_CERT='%s'" % master_ssl_cert)
        if master_ssl_key:
            chm.append("MASTER_SSL_KEY='%s'" % master_ssl_key)
        if master_ssl_cipher:
            chm.append("MASTER_SSL_CIPHER='%s'" % master_ssl_cipher)
        if master_auto_position:
            chm.append("MASTER_AUTO_POSITION=1")
        if master_use_gtid is not None:
            chm.append("MASTER_USE_GTID=%s" % master_use_gtid)
        try:
            changemaster(cursor, chm, connection_name, channel)
        except mysql_driver.Warning as e:
            result["warning"] = to_native(e)
        except Exception as e:
            module.fail_json(msg="%s. Query == CHANGE MASTER TO %s" % (to_native(e), chm))
        result["changed"] = True
        module.exit_json(queries=executed_queries, **result)
    elif mode in "startslave":
        started = start_slave(cursor, connection_name, channel)
        if started is True:
            module.exit_json(msg="Slave started ", changed=True, queries=executed_queries)
        else:
            module.exit_json(msg="Slave already started (Or cannot be started)", changed=False, queries=executed_queries)
    elif mode in "stopslave":
        stopped = stop_slave(cursor, connection_name, channel)
        if stopped is True:
            module.exit_json(msg="Slave stopped", changed=True, queries=executed_queries)
        else:
            module.exit_json(msg="Slave already stopped", changed=False, queries=executed_queries)
    elif mode in "resetslave":
        reset = reset_slave(cursor, connection_name, channel)
        if reset is True:
            module.exit_json(msg="Slave reset", changed=True, queries=executed_queries)
        else:
            module.exit_json(msg="Slave already reset", changed=False, queries=executed_queries)
    elif mode in "resetslaveall":
        reset = reset_slave_all(cursor, connection_name, channel)
        if reset is True:
            module.exit_json(msg="Slave reset", changed=True, queries=executed_queries)
        else:
            module.exit_json(msg="Slave already reset", changed=False, queries=executed_queries)

    warnings.simplefilter("ignore")


if __name__ == "__main__":
    main()
